File: Lab-12/conVolParalela.java
import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.LinkedList;
import java.util.List;
import java.util.Random;
import java.util.Scanner;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

/**
 * conVolParalela
 * 
 * @author Nicolas Ruiz Requejo
 */

public class conVolParalela implements Runnable {

    private static final int N = 1000;
    private static final int M = 1000;

    public static final int[][] enfoque = { { 0, -1, 0 }, { -1, 5, -1 }, { 0, -1, 0 } };
    public static final int[][] realzaBordes = { { 0, 0, 0 }, { -1, 1, 0 }, { 0, 0, 0 } };
    public static final int[][] detectaBordes = { { 0, 1, 0 }, { 1, -4, 1 }, { 0, 1, 0 } };
    public static final int[][] sobel = { { -1, 0, 1 }, { -2, 0, 2 }, { -1, 0, 1 } };
    public static final int[][] sharpen = { { 1, -2, 1 }, { -2, 5, -2 }, { 1, -2, 1 } };

    public static final int[][] prueba = { { 35, 40, 41, 45, 50 }, { 40, 40, 42, 46, 52 }, { 42, 46, 50, 55, 55 },
            { 48, 52, 56, 58, 60 }, { 56, 60, 65, 70, 75 } };
    public static final int[][] kernelprueba = { { -2, -1, 0 }, { -1, 1, 1 }, { 0, 1, 2 } };

    public static int[][] A = new int[N][M];
    public static int[][] C = new int[N][M];

    private int ini, fini;
    private int[][] mat, h;

    public conVolParalela(int ini, int fini, int[][] mat, int[][] kernel) {
        this.ini = ini;
        this.fini = fini;
        this.mat = mat;
        h = kernel;
    }

    @Override
    public void run() {
        for (int i = ini; i <= fini; i++) {
            for (int j = 1; j < (M - 1); j++) {
                for (int rowoffset = -1; rowoffset <= 1; rowoffset++) {
                    for (int coloffset = -1; coloffset <= 1; coloffset++) {
                        C[i][j] += mat[i + rowoffset][j + coloffset] * h[1 + rowoffset][1 + coloffset];
                    }
                }
            }
        }
    }

    public static void main(String[] args) throws InterruptedException, IOException {
        Scanner sc = new Scanner(System.in);

        System.out.println("Numero de tareas: ");
        int numHilos = sc.nextInt();

        System.out.println("1. Enfoque");
        System.out.println("2. Realzar bordes");
        System.out.println("3. Detectar bordes");
        System.out.println("4. Sobel");
        System.out.println("5. Sharpen");
        System.out.println("6. Prueba");
        System.out.println("Elija un kernel: ");
        int op = sc.nextInt();
        int[][] kernel = null;
        switch (op) {
            case 1:
                kernel = conVolParalela.enfoque;
                break;
            case 2:
                kernel = conVolParalela.realzaBordes;
                break;
            case 3:
                kernel = conVolParalela.detectaBordes;
                break;
            case 4:
                kernel = conVolParalela.sobel;
                break;
            case 5:
                kernel = conVolParalela.sharpen;
                break;
            case 6:
                kernel = conVolParalela.kernelprueba;
            default:
                break;
        }

        rellenoAleatorio(A);

        long initime = System.nanoTime();
        // reparto
        List<Runnable> tareas = new LinkedList<Runnable>();
        int numTareas = (N - 2) / numHilos;
        int restoTareas = (N - 2) % numHilos;
        int i;
        int n = N - 2 - restoTareas; // filas evaluadas
        //System.out.println("numTareas = " + numTareas);
        //System.out.println("resto = " + restoTareas);
        //System.out.println("n = " + n);
        for (i = 1; i <= (n - numTareas); i += numTareas) {
            //System.out.println("ini= " + i + " fin= " + ((i + numTareas) - 1));
            tareas.add(new conVolParalela(i, (i + numTareas) - 1, A, kernel));
        }
        if (restoTareas != 0) {
            //System.out.println("RESTO ini= " + i + " fin= " + ((i + numTareas + restoTareas) - 1));
            tareas.add(new conVolParalela(i, (i + numTareas + restoTareas) - 1, A, kernel));
        } else {
            //System.out.println("EXACTO ini= " + i + " fin= " + ((i + numTareas) - 1));
            tareas.add(new conVolParalela(i, (i + numTareas) - 1, A, kernel));
        }

        ExecutorService pool = Executors.newFixedThreadPool(numHilos);
        for (Runnable runnable : tareas) {
            pool.execute(runnable);
        }

        pool.shutdown();
        pool.awaitTermination(10, TimeUnit.DAYS);

        long endtime = System.nanoTime();
        double exetime = ((double) endtime - initime) / 1.0e6;

        System.out.println("Tiempo de ejecucion: " + exetime + " milisegundos.");
        //printMatrix(prueba, "orig");
        //printMatrix(C, "resultado");
    }

    private static void rellenoAleatorio(int[][] mat) {
        Random ran = new Random(System.nanoTime());

        for (int i = 0; i < mat.length; ++i)
            for (int j = 0; j < mat[0].length; ++j)
                mat[i][j] = ran.nextInt(255);
    }

    private static void printMatrix(int[][] mat, String name) throws IOException {
        StringBuilder build = new StringBuilder();
        int rows = mat.length;
        int cols = mat[0].length;
        build.append("[");
        for (int i = 0; i < rows; ++i) {
            for (int j = 0; j < cols; ++j) {
                build.append(mat[i][j] + " ");
            }
            build.append('\n');
        }
        build.append("]");

        BufferedWriter writer = new BufferedWriter(new FileWriter(name + ".txt"));
        writer.write(build.toString());// save the string representation of the board
        writer.close();
    }

}

File: Lab-12/conVolParalela.cpp
/**
 *  conVolParalela
 * 
 *  @author Nicolas Ruiz Requejo
 **/

#include <iostream>
#include <random>
#include <chrono>
#include <thread>
#include <vector>
#include <algorithm>

#define N 1000
#define M 1000

void rellenoAleatorio(int mat[][M], int dimN, int dimM)
{
    unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
    std::default_random_engine generator(seed);

    std::uniform_int_distribution<int> distribution(1, 100);

    for (int i = 0; i < dimN; ++i)
        for (int j = 0; j < dimM; ++j)
            mat[i][j] = distribution(generator);
}

void printMatrix(int mat[][M], int dimN, int dimM)
{
    std::cout << "[";
    for (size_t i = 0; i < dimN; i++)
    {
        for (size_t j = 0; j < dimM; j++)
        {
            std::cout << mat[i][j] << " ";
        }
        std::cout << std::endl;
    }
    std::cout << "]" << std::endl;
}

void copy2DArray(int src[3][3], int dst[3][3])
{
    for (size_t i = 0; i < 3; i++)
    {
        for (size_t j = 0; j < 3; j++)
        {
            dst[i][j] = src[i][j];
        }
    }
}

int enfoque[][3] = {{0, -1, 0}, {-1, 5, -1}, {0, -1, 0}};
int realzaBordes[][3] = {{0, 0, 0}, {-1, 1, 0}, {0, 0, 0}};
int detectaBordes[][3] = {{0, 1, 0}, {1, -4, 1}, {0, 1, 0}};
int sobel[][3] = {{-1, 0, 1}, {-2, 0, 2}, {-1, 0, 1}};
int sharpen[][3] = {{1, -2, 1}, {-2, 5, -2}, {1, -2, 1}};
int prueba[][5] = {{35, 40, 41, 45, 50}, {40, 40, 42, 46, 52}, {42, 46, 50, 55, 55}, {48, 52, 56, 58, 60}, {56, 60, 65, 70, 75}};
int kernelprueba[][3] = {{-2, -1, 0}, {-1, 1, 1}, {0, 1, 2}};

int A[N][M];
int h[3][3];
static int C[N][M];

class conVolParalela
{
private:
    int ini, fini;

public:
    conVolParalela(int ini, int fini)
        : ini(ini), fini(fini){};
    void operator()() const
    {
        for (int i = ini; i <= fini; i++)
        {
            for (int j = 1; j < (M - 1); j++)
            {
                for (int rowoffset = -1; rowoffset <= 1; rowoffset++)
                {
                    for (int coloffset = -1; coloffset <= 1; coloffset++)
                    {
                        C[i][j] += A[i + rowoffset][j + coloffset] * h[1 + rowoffset][1 + coloffset];
                    }
                }
            }
        }
    };
};

int main(int argc, char const *argv[])
{
    using namespace std::chrono;

    std::cout << "Numero de hilos: ";
    int numHilos;
    std::cin >> numHilos;

    std::cout << "1. Enfoque" << std::endl;
    std::cout << "2. Realzar bordes" << std::endl;
    std::cout << "3. Detectar bordes" << std::endl;
    std::cout << "4. Sobel" << std::endl;
    std::cout << "5. Sharpen" << std::endl;
    std::cout << "6. Prueba" << std::endl;
    std::cout << "Elija un kernel: " << std::endl;
    int op = 0;
    std::cin >> op;

    switch (op)
    {
    case 1:
        copy2DArray(enfoque, h);
        break;
    case 2:
        copy2DArray(realzaBordes, h);
        break;
    case 3:
        copy2DArray(detectaBordes, h);
        break;
    case 4:
        copy2DArray(sobel, h);
        break;
    case 5:
        copy2DArray(sharpen, h);
        break;
    case 6:
        copy2DArray(kernelprueba, h);
        break;
    default:
        break;
    }

    rellenoAleatorio(A, N, M);

    high_resolution_clock::time_point t1 = high_resolution_clock::now();
    //reparto
    std::vector<conVolParalela> tareas;
    int numTareas = (N - 2) / numHilos;
    int restoTareas = (N - 2) % numHilos;
    int i;
    int n = N - 2 - restoTareas; // filas evaluadas
    //std::cout << "numTareas = " << numTareas << std::endl;
    //std::cout << "resto = " << restoTareas << std::endl;
    //std::cout << "n = " << n << std::endl;
    for (i = 1; i <= (n - numTareas); i += numTareas)
    {
        //std::cout << "ini= " << i << " fin= " << ((i + numTareas) - 1) << std::endl;
        tareas.push_back(conVolParalela(i, (i + numTareas) - 1));
    }
    if (restoTareas != 0)
    {
        //std::cout << "RESTO ini= " << i << " fin= " << ((i + numTareas + restoTareas) - 1) << std::endl;
        tareas.push_back(conVolParalela(i, (i + numTareas + restoTareas) - 1));
    }
    else
    {
        //std::cout << "EXACTO ini= " << i << " fin= " << ((i + numTareas) - 1) << std::endl;
        tareas.push_back(conVolParalela(i, (i + numTareas) - 1));
    }

    // co-rutina
    std::vector<std::thread> hilos;
    for (conVolParalela &t : tareas)
    {
        hilos.push_back(std::thread(t));
    }

    for (std::thread &t : hilos)
    {
        t.join();
    }

    high_resolution_clock::time_point t2 = high_resolution_clock::now();
    duration<double, std::milli> time_span = duration_cast<duration<double, std::milli>>(t2 - t1);

    std::cout << "Tiempo de ejecucion: " << time_span.count() << " milisegundos." << std::endl;

    return 0;
}
